File: 217. Contains Duplicate.cpp
#include <iostream>
#include <map>
#include <vector>
using namespace std;

class Solution {
public:
    // // sort: O(NlogN) time without extra space, 91.01%, ok
     bool containsDuplicate(vector<int>& nums) {
        if(nums.empty()) return false;
        sort(nums.begin(), nums.end());
        for(int i = 1; i < nums.size(); i++)
            if(nums[i] == nums[i - 1])
                return true;
        return false;
     }

    // hashTable: O(N) time and space, 71.78%, ok
    bool containsDuplicate(vector<int>& nums) {
        unordered_map<int, int> flag;
        for(int i = 0; i < nums.size(); i++)
            if(flag.count(nums[i])) return true;
            else flag[nums[i]] = 1;
        return false;
    }

    /*  For certain test cases with not very large N, the runtime of hashtable method can be slower than sort method.
        The reason is hash table has some overhead in maintaining its property.
        Real world performance can be different from what the Big-O notation says.
        The Big-O notation only tells us that for sufficiently large input, one will be faster than the other.
        Therefore, when N is not sufficiently large, an O(n) algorithm can be slower than an O(nlogn) algorithm. */
};

int main(){
	int testcase[] = {1,2,3,4,5,1};
	vector<int> test;
	for(int i = 0; i < sizeof(testcase) / sizeof(int); i ++){
		test.push_back(testcase[i]);
	}
	if(containsDuplicate(test))
		cout << "true";
}

File: 217. Contains Duplicate.py
class Solution(object):
    
    # brute force: O(N2) time, O(1) space, TLE
    def containsDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        for i in xrange(len(nums)):
            for j in xrange(i + 1, len(nums)):
                if nums[i] == nums[j]:
                    return True
        return False
    
    # hash table: O(N) time & space
    def containsDuplicate(self, nums):
        count = {}
        for num in nums:
            if num in count:
                return True
            count[num] = 1
        return False
    
    # hash set: O(N) time & space
    def containsDuplicate(self, nums):
        count = set(nums)
        return len(count) != len(nums)
    
    # sort: O(NlogN) time, O(1) space
    def containsDuplicate(self, nums):
        nums.sort()
        prev = None
        for num in nums:
            if prev is not None and num == prev:
                return True
            prev = num
        return False
        